### ISSUE TITLE
why share the same value of class member when using multigpu of HybridBlock, bug?

### ISSUE THREAD

**Issue by tornadomeet** [apache/incubator-mxnet#9338]
## Description
in my opinion, when using multi-gpu for training, each device should do forward() on one device(gpu) independently, and it should use it's own class member value, not share class member value cross devices. Is this correct?  

but i found when using HybridBlock, multi devices share the same value, why?

ubuntu16.04
python2.7.13

MXNet commit hash:
```deb25bca9803cac338faedb2eb14dfda22813b75```

## Minimum reproducible example
```python
# mnist_bug.py
import argparse,logging
import numpy as np
import mxnet as mx
from mxnet import gluon, autograd
from mxnet.gluon import nn

logging.basicConfig(level=logging.DEBUG)
batch_size = 128
epochs = 10

class BugBlock(gluon.HybridBlock):
    def __init__(self):
        super(BugBlock, self).__init__()
        self.cnt = 0
    def hybrid_forward(self, F, x):
        self.cnt += 1
        print("ctx={}\tcnt={}".format(x.context, self.cnt))
        return x

net = nn.HybridSequential()
with net.name_scope():
    net.add(nn.Dense(1280, activation='relu'))
    net.add(nn.Dense(640, activation='relu'))
    net.add(nn.Dense(10))
    net.add(BugBlock())

def transformer(data, label):
    data = data.reshape((-1,)).astype(np.float32)/255
    return data, label

train_data = gluon.data.DataLoader(
    gluon.data.vision.MNIST('./data', train=True, transform=transformer),
    batch_size=batch_size, shuffle=True, last_batch='discard')

def train(epochs, ctx):
    net.initialize(ctx=ctx)
    trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': 0.1, 'momentum': 0.9})
    metric = mx.metric.Accuracy()
    loss = gluon.loss.SoftmaxCrossEntropyLoss()
    for epoch in range(epochs):
        metric.reset()
        for i, (data, label) in enumerate(train_data):
            data = gluon.utils.split_and_load(data, ctx_list=ctx, batch_axis=0)
            label = gluon.utils.split_and_load(label, ctx_list=ctx, batch_axis=0)
            outputs = []
            Ls = []
            with autograd.record():
                for x, y in zip(data, label):
                    z = net(x)
                    L = loss(z, y)
                    Ls.append(L)
                    outputs.append(z)
                for L in Ls:
                    L.backward()
            trainer.step(data[0].shape[0])
            metric.update(label, outputs)

if __name__ == '__main__':
    ctx = [mx.gpu(0), mx.gpu(1)]
    train(epochs, ctx)

```

## Steps to reproduce
run ```python mnist_bug.py```, outputs log will like this:

![image](https://user-images.githubusercontent.com/5860892/34650916-cea30994-f403-11e7-8f0a-0e83373be994.png)

which `cnt` is share like a static variable.

@szha @piiswrong

**Comment by szha** [comment]
Regular instance variables behave like instance variables, agnostic of the context.
Parameter instance variables will have a copy for each of the context and behave as you stated.

Is there a use case where you need context-dependent copy of regular variables?

**Comment by tornadomeet** [comment]
hello, @szha ，` self.cnt` is parameter instance variable or regular instance variabel.

there are may case which using it own class member value independently, such as 
```python
class ToyBlock(gluon.HybridBlock):
    """just a toy block for explain inint class member for each device"""
    def __init__(self):
        super(BugBlock, self).__init__()

    def initialize(self, features, F=mx.nd):
        self._initialized = True 
        if F == mx.nd:
            idx_vec = F.arange(0, stop=self._N, ctx=features.context)
        elif F == mx.sym:
            idx_vec = F.arange(0, stop=self._N)
        self._eye_matrix = F.stop_gradient(F.broadcast_equal(F.expand_dims(idx_vec, 0), F.expand_dims(idx_vec, 1)))

    def hybrid_forward(self, F, x):
        if not self._initialized:
            self.initialize(features, F)
        loss = F.mean(F.sqrt(F.maximum(1e-10, 2.0 - 2.0 * F.dot(features, features, transpose_b=True))) - self._eye_matrix)
        return loss
```
because we want only init `self._eye_matirx` only one time, so when `F=mx.nd`, when should know the context advance for initialized.(of course if whe initialize every time when do forward(), then it will be ok).

but when using multi-gpu for training, after forward on gpu0, `self._initialized=True`, then  gpu1 will not initialize it's own `self._eye_matrix `.

**Comment by szha** [comment]
`cnt` would be a regular variable because `self.cnt = 0`.
In the eye case what you want is a parameter that remains constant. You can achieve that by using `self.params.get('eye_matrix', grad_req='null')`.

**Comment by szha** [comment]
I recall that Haibin looked into 'eye' a while back. @eric-haibin-lin do you have recommendation on how to implement the above case?

**Comment by tornadomeet** [comment]
yes, `eye_matrix` is just one example, what i need is a constant `ndarray` with specific context when F=mx.nd;

another, if using `mx.operator.CustomOp` for implement user-defined operator, `self.cnt` will be  hold by each device independently, so i need the same behavior in HybridBlock.

**Comment by karan6181** [comment]
@sandeep-krishnamurthy Can you please add labels:
- Gluon
- Python

Thanks!
